lib/code-callers.js

```diff
@@ -27,39 +30,72 @@ Load is reported in:
 
 */
 
-let pythonCallers = []; // all pythonCallers, whether available for work or currently busy
-let availablePythonCallers = []; // FIFO queue for available pythonCallers waiting for a job/callback
-let waitingCallbacks = []; // FIFO queue for jobs/callbacks waiting for an available pythonCaller
+/** @typedef {PythonCaller | DockerCaller } CodeCaller */
+
+/**
+ * All pythonCallers, whether available for work or currently busy
+ *
+ * @type {CodeCaller[]}
+ */
+let pythonCallers = [];
+
+/**
+ * FIFO queue for available pythonCallers waiting for a job/callback
+ *
+ * @type {CodeCaller[]}
+ */
+let availablePythonCallers = [];
+
+/** @typedef {{ (err: Error | null, caller: CodeCaller): void, __load_uuid: string }} InternalCallback */
+
+/**
+ * FIFO queue for jobs/callbacks waiting for an available pythonCaller
+ * @type {Array<[string, InternalCallback]>}
+ */
+let waitingCallbacks = [];
+
+// Will be set to the appropriate class in init()
+let CodeCaller = null;
+
+let cancelWarmup = null;
 
 module.exports = {
   init() {
     debug('init()');
-    if (!config.useWorkers) {
-      return;
+    const { workersExecutionMode } = config;
+    if (workersExecutionMode === 'container') {
+      CodeCaller = DockerCaller;
+    } else if (workersExecutionMode === 'native') {
+      CodeCaller = PythonCaller;
+    } else {
+      throw new Error(`unknown config.workersExecutionMode: ${workersExecutionMode}`);
     }
+
     var numWorkers = config.workersCount;
     if (numWorkers == null) {
       numWorkers = Math.ceil(config.workersPerCpu * os.cpus().length);
     }
+
     for (let i = 0; i < numWorkers; i++) {
-      const pc = new codeCaller.PythonCaller();
-      pc.number = i;
+      const pc = new CodeCaller({
+        questionTimeoutMilliseconds: config.questionTimeoutMilliseconds,
+      });
       pythonCallers.push(pc);
+      load.startJob('python', pc.uuid);
       load.startJob('python_worker_idle', pc.uuid);
       availablePythonCallers.push(pc);
     }
-    module.exports._warmUpWorkers();
+    cancelWarmup = module.exports._warmUpWorkers();
```

Comment:
I introduced this so that we could cancel warming up the code callers in `finish()` - AFAICT, the fact that we were *not* doing that before was a bug, since in tests we could call `finish()` before workers had finished warming up, which would warm up workers that no longer existed.